### PR TITLE
[mlir][vector] ND vectors linearization pass

### PR DESCRIPTION
**Pull request by Hardcode84**
Common backends (LLVM, SPIR-V) only supports 1D vectors, LLVM conversion handles ND vectors (N >= 2) as `array<array<... vector>>` and SPIR-V conversion doesn't handle them at all at the moment. Sometimes it's preferable to treat multidim vectors as linearized 1D. Add pass to do this. Only constants and simple elementwise ops are supported for now.

@krzysz00 I've extracted yours result type conversion code from LegalizeToF32 and moved it to common place.

Also, add ConversionPattern class operating on traits.

### CODE REVIEW

**Review by dcaballe — CHANGES_REQUESTED** [review]
Hey, thanks for the contribution!

I think we should look at applying vector unrolling instead or linearization. It's currently happening at LLVM level but we have always talked about moving it to an earlier stage so that the lowering to LLVM and SPIR-V are aligned. 

Collapsing some vector dimensions is definitely beneficial for some cases, esp. when the trailing dimension can't fill a full 1-D physical vector register, such as in your test (I think @hanhanW was working on something for these cases). However, I'm not sure we can apply it in a general way. For example, we couldn't turn an arbitrary `vector<2x2xf32>` load/store into a `vector<4xf32>` one because elements might not be contiguous in memory across the dimensions. Also, by linearizing an n-D vector instead of unrolling it, we are losing control on the actual vector length used and relying on the hardware backend to do what we expected, which is something we have been trying to avoid.

Would vector unrolling work for your case?

**Comment by Hardcode84** [comment]
In our case underlying API/ABI (SPIR-V intrinsics) expects 1D vectors, but it interprets them as ND and we represent them (and surrounding arith ops) as ND at the intermediate steps, so we need to linearize it at some point. Also, vector unrolling and linearization are not mutually exclusive. We can have some partial unroll first and than run linearization on result (and potentially, some legalization pass after, breaking big 1D vectors into smaller ones, supported by HW). Regarding loads/stores and similar potentially non-linearizable ops, they are not covered by the patch and will stay untouched + `shape_cast`, but we can unroll them and reconstruct linear vector if we cannot statically determine they are safe.

I understand this transform is quite usecase-specific, so don't want to force it into default vector/llvm/spir-v pipeline.

**Comment by dcaballe** [comment]
Ok, got it. Would it make sense then to remove the pass in `Vector/Transforms` and just expose this transformation through the `populate...` API? We follow this approach for cases like this. You can move that pass under `mlir/test/lib/Dialect/Vector` for testing. In that way, the pass is not prescribing how the patterns should be used and leave that decision to the user.

**Comment by dcaballe** [comment]
Perhaps we can also add a callback function so that the user can decide when an op should be linearized or not. That should give us the level of flexibility that we need.

**Comment by Hardcode84** [comment]
> You can move that pass under mlir/test/lib/Dialect/Vector for testing. In that way, the pass is not prescribing how the patterns should be used and leave that decision to the user.

Makes sense, I'll do it.

> Perhaps we can also add a callback function so that the user can decide when an op should be linearized or not. That should give us the level of flexibility that we need.

User already have some control over it by overriding `target.mark(Unknown)OpDynamicallyLegal`.

**Comment by rengolin** [comment]
Not directly applicable, but [our solution](https://github.com/plaidml/tpp-mlir/blob/9d5f72003cae98468689c829bef33894501237eb/tools/tpp-run/MLIRBench.cpp#L347) to a similar problem in memref was to linearize all but the last dimensions (for D>2).

This would keep the inner dimension intact and not make assumptions about the vector length, but would potentially break noncontiguous higher dimensions if the layout isn't changed accordingly?

But this does require the underlying user (print in out case) to change behaviour, which is not always a generic property.

File: mlir/test/Dialect/Vector/linearize.mlir [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// RUN: mlir-opt %s -split-input-file -test-vector-linearize | FileCheck %s
+
+// CHECK-LABEL: test_linearize
+//  CHECK-SAME: (%[[ORIG_ARG:.*]]: vector<2x2xf32>)
+//       CHECK: %[[ARG:.*]] = vector.shape_cast %[[ORIG_ARG]] : vector<2x2xf32> to vector<4xf32>
+func.func @test_linearize(%arg0: vector<2x2xf32>) -> vector<2x2xf32> {
+//       CHECK: %[[C1:.*]] = arith.constant dense<[1.000000e+00, 2.000000e+00, 3.000000e+00, 4.000000e+00]> : vector<4xf32>
+  %0 = arith.constant dense<[[1.0, 2.0], [3.0, 4.0]]> : vector<2x2xf32>
+// Arith and math ops are handled in generic way, check some of them
+//       CHECK: %{{.*}} =  math.sin %[[ARG]] : vector<4xf32>
```
Comment (dcaballe) [inline mlir/test/Dialect/Vector/linearize.mlir:10]: Could we add a few more tests? Perhaps some where the transformation shouldn't trigger?

File: mlir/include/mlir/Dialect/Vector/Transforms/VectorRewritePatterns.h [hunk 2]
```diff
@@ -375,6 +377,11 @@ void populateVectorNarrowTypeRewritePatterns(RewritePatternSet &patterns,
 void populateVectorTransposeNarrowTypeRewritePatterns(
     RewritePatternSet &patterns, PatternBenefit benefit = 1);
 
+/// Linearizes ND vectors (N >= 2) into 1D.
+void populateVectorLinearizeTypeConversionsAndLegality(
```
[the remark this answers is not in the document]
Comment (Hardcode84) [reply mlir/include/mlir/Dialect/Vector/Transforms/VectorRewritePatterns.h]: Updated the doc

File: mlir/test/Dialect/Vector/linearize.mlir (hunk 1 above)
Comment (Hardcode84) [reply mlir/test/Dialect/Vector/linearize.mlir:12]: Added more checks for `shape_cast` on the boundaries and check for `return` op unchanged, not sure what else to check.

**Review by dcaballe — APPROVED** [review]
LG, thanks!

**Pull request merged**